File: code/mathematical-algorithms/catalan_number/catalan_number.rb
def catalan(num)
  return 1 if num <= 1
  ans = 0
  i = 0

  while i < num
    first = catalan i
    second = catalan num - i - 1
    ans += (first * second)
    i += 1
  end
  ans
end

Integer x = 1
while x <= 10
  res = catalan x
  puts res
  x += 1
end

File: code/mathematical-algorithms/catalan_number/catalan_number.go
package main

import "fmt"

func main() {
	for i := 0; i < 10; i++ {
		fmt.Println(catalan(i))
	}
}

func catalan(i int) int {
	coeff := binomialCoeff(2*i, i)
	return coeff / (i + 1)
}

func binomialCoeff(n, k int) int {
	res := 1

	if k > n-k {
		k = n - k
	}

	for j := 0; j < k; j++ {
		res *= (n - j)
		res /= (j + 1)
	}

	return res
}
